modules/dag-l0/src/main/scala/org/tessellation/dag/l0/domain/snapshot/programs/GlobalSnapshotEventCutter.scala

```diff
@@ -0,0 +1,144 @@
+          case None => None
+          case Some(globalPeerIdEvent) =>
+            val globalSnapshotEvent = eventToRemove.toGlobalSnapshotEvent
+            val updatedEvents = globalPeerIdEvent.filterNot { case (_, gse) => gse == globalSnapshotEvent }
+            Option.when(updatedEvents.nonEmpty)(updatedEvents)
+        }
+
+      trait PeerIdEvent {
+        def id: PeerId
+        def fee: Long
+        def hash: Hash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent
+      }
+
+      val ordering: Ordering[PeerIdEvent] = Ordering.by[PeerIdEvent, (Long, Hash)](e => (e.fee, e.hash))
+
+      case class PeerIdStateChannelEvent(id: PeerId, event: StateChannelEvent) extends PeerIdEvent {
+        val fee: Long = event.snapshotBinary.value.fee.value.value
+        def hash: Hash = event.snapshotBinary.lastSnapshotHash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent = event.asLeft[DAGEvent]
+      }
+
+      case class PeerIdHashedEvent(id: PeerId, event: Hashed[Block]) extends PeerIdEvent {
+        val fee: Long = event.signed.value.transactions.toIterable.map(_.value.fee.value.value).sum
+        def hash: Hash = event.hash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent = event.signed.asRight[StateChannelEvent]
+
+        def isMyChild(that: PeerIdHashedEvent): Boolean =
+          that.event.signed.value.parent.contains_(event.ownReference)
+      }
+
+      class OrderedEvents(sortedStateEvents: List[PeerIdStateChannelEvent], hashedEvents: List[PeerIdHashedEvent]) {
+        def next: Option[(PeerIdEvent, OrderedEvents)] = {
+          val (parents, nonParents) =
+            hashedEvents.foldLeft((List.empty[PeerIdHashedEvent], List.empty[PeerIdHashedEvent])) {
+              case ((parents, nonParents), p) if hashedEvents.exists(p.isMyChild) => (p :: parents, nonParents)
+              case ((parents, nonParents), p)                                     => (parents, p :: nonParents)
+            }
+
+          (sortedStateEvents, nonParents.sorted(ordering)) match {
+            case (Nil, Nil) =>
+              none
+            case (h :: t, Nil) =>
+              (h, new OrderedEvents(t, Nil)).some
+            case (Nil, h :: t) =>
+              (h, new OrderedEvents(Nil, t ++ parents)).some
+            case (h1 :: t1, h2 :: t2) =>
+              if (h1.fee < h2.fee || (h1.fee == h2.fee && sortedStateEvents.length >= hashedEvents.length))
+                (h1, new OrderedEvents(t1, hashedEvents)).some
+              else
+                (h2, new OrderedEvents(sortedStateEvents, t2 ++ parents)).some
+          }
+        }
+      }
+
+      object OrderedEvents {
+        // NOTE: Order for removal: NOT parent → Fee → Split by event type → Hash
+        def apply(events: GlobalPeerIdEventMap): F[OrderedEvents] =
+          unpackGlobalPeerIdEvents(events).map {
+            case (stateChannelEvents, hashedEvents) =>
+              new OrderedEvents(stateChannelEvents.sorted(ordering), hashedEvents)
```

Comment:
I don't think `hashedEvents` are ordered at this point (also there is no total ordering on block I think)

Reply:
@TheMMaciek I thought of that and checked, and they are still ordered. However, it did bother me that this may not be something we can rely on. I'll look into it again.